// o1js/dist/node/lib/util/global-context.js
export { Context };
const Context = { create };
function create(options = {
    allowsNesting: true,
    default: undefined,
}) {
    let t = Object.assign(function () {
        return t.data[t.data.length - 1]?.context;
    }, {
        data: [],
        allowsNesting: options.allowsNesting ?? true,
        get: () => get(t),
        has: () => t.data.length !== 0,
        runWith(context, func) {
            let id = enter(t, context);
            let result;
            let resultContext;
            try {
                result = func(context);
            }
            finally {
                resultContext = leave(t, id);
            }
            return [resultContext, result];
        },
        async runWithAsync(context, func) {
            let id = enter(t, context);
            let result;
            let resultContext;
            try {
                result = await func(context);
            }
            finally {
                resultContext = leave(t, id);
            }
            return [resultContext, result];
        },
        enter: (context) => enter(t, context),
        leave: (id) => leave(t, id),
        id: () => {
            if (t.data.length === 0)
                throw Error(contextConflictMessage);
            return t.data[t.data.length - 1].id;
        },
    });
    if (options.default !== undefined)
        enter(t, options.default);
    return t;
}
function enter(t, context) {
    if (t.data.length > 0 && !t.allowsNesting) {
        throw Error(contextConflictMessage);
    }
    let id = Math.random();
    t.data.push({ context, id });
    return id;
}
function leave(t, id) {
    let current = t.data.pop();
    if (current === undefined)
        throw Error(contextConflictMessage);
    if (current.id !== id)
        throw Error(contextConflictMessage);
    return current.context;
}
function get(t) {
    if (t.data.length === 0)
        throw Error(contextConflictMessage);
    let current = t.data[t.data.length - 1];
    return current.context;
}
// FIXME there are many common scenarios where this error occurs, which weren't expected when this was written
// it should list them and help to resolve them
let contextConflictMessage = "It seems you're running multiple provers concurrently within" +
    ' the same JavaScript thread, which, at the moment, is not supported and would lead to bugs.';
